### ISSUE TITLE
Vault roleset update causes stale Service Account bindings.

### ISSUE THREAD

**Issue by sireesha-sura** [hashicorp/vault-plugin-secrets-gcp#95]
Opening an issue as suggested by our Google PSO.

Our internal GCP project provisioning tool uses a Vault roleset to create new GCP projects and take further actions on the project (ex., fetch CloudSQL instances in the project). Pared down version of the roleset:

`
{
  "project": "admin-project",
  "secret_type": "access_token",
  "token_scopes": "https://www.googleapis.com/auth/userinfo.email https://www.googleapis.com/auth/cloud-platform",
  "bindings": "resource \"//cloudresourcemanager.googleapis.com/organizations/<orgId>\" {roles = [\"roles/resourcemanager.projectCreator\", \"roles/resourcemanager.projectDeleter\", ...]}"
  ...
}
`

The roleset is granted projectCreator role at the Organization level; for the rest of the IAM permissions pertaining to the GCP project's resources, the tool relies on the fact that the service account was used to create the GCP project and hence has role/owner on the project.

On 8/19, we updated our Vault roleset with additional permissions. This resulted in the backing service account name to be updated (from "vaultroleset-1583368845@admin-project.iam.gserviceaccount.com" to "vaultroleset-1598036538@admin-project.iam.gserviceaccount.com").

However, the new Service account name was not propagated to all projects that were created before the roleset update. The older projects continued to have stale IAM bindings with subject as "vaultroleset-1583368845@admin-project.iam.gserviceaccount.com". This caused the tool to fail with 403 errors when attempting to fetch resources in the project.

A similar roleset update back in March did not cause any issues.

**Comment by technologik** [comment]
We just encountered a similar issue again today. Here's the process we identified that caused us to get in this state:

1. An individual IAM permission in GCP (in this case `role/pubsub.admin`) on a GCP SA managed in the Vault GCP Secrets backened is removed out-of-band (ie - it was removed in the GCP console, not with the Vault roleset config). This is an anti-pattern that shouldn't happen. It was a mistake. But it happened all the same.

2. When this was identified we tried re-writing the roleset:  

```
curl --header "X-Vault-Token: <TOKEN>" --request POST --data @path/to/roleset/config https://<vault-dns>:8200/v1/gcp/roleset/<roleset-name>
```

This had no effect. It did not add the IAM binding in GCP.

3. removing the permission from the roleset, and adding it back in again deletes the entire GCP SA and makes a new service account. At which point any references to the existing SA (say, as infrastructure as code) break.
